src/Stomp/Network/Observer/Heartbeat/Emitter.php

```diff
@@ -0,0 +1,270 @@
+     *
+     * @return void
+     */
+    public function emptyLineReceived()
+    {
+        $this->onPassiveEvent();
+    }
+
+    /**
+     * An passive event was fired, all that triggers a connection activity that is not leading to outgoing traffic.
+     *
+     * @return void
+     */
+    private function onPassiveEvent()
+    {
+        if ($this->isDelayed()) {
+            $this->sendBeat();
+        }
+    }
+
+    /**
+     * Check if the emitter is in a state that indicates a delay.
+     *
+     * @return bool
+     */
+    public function isDelayed()
+    {
+        if ($this->enabled) {
+            $now = microtime(true);
+            return ($now - $this->lastbeat > $this->interval);
+        }
+        return false;
+    }
+
+    /**
+     * Send a beat to the server.
+     *
+     * @return void
+     */
+    private function sendBeat()
+    {
+        $this->connection->sendAlive();
+        $this->notifyBeat();
+    }
+
+    /**
+     * Outgoing activity event.
+     *
+     * @return void
+     */
+    private function notifyBeat()
+    {
+        $this->lastbeat = microtime(true);
+    }
+
+    /**
+     * Indicates that during a read call no frame was received, but an EOL line.
+     *
+     * @return void
+     */
+    public function emptyBuffer()
+    {
+        $this->onPassiveEvent();
+    }
+
+    /**
+     * Indicates that a frame has been received from the server.
+     *
+     * @param Frame $frame that has been received
+     * @return void
+     */
+    public function receivedFrame(Frame $frame)
+    {
+        if ($frame->getCommand() === Emitter::FRAME_SERVER_CONNECTED) {
+            $beats = $this->getHeartbeats($frame);
+            $this->intervalServer = $beats[1];
+            if ($this->intervalServer && ($this->intervalClient || $this->intervalClient === null)) {
+                $this->interval = $this->intervalUsage * (max($this->intervalServer,
+                            $this->intervalClient) / 1000); // milli to micro
+                $this->enabled = true;
+            }
+        } else {
+            $this->onPassiveEvent();
+        }
+    }
+
+    /**
+     * Returns the heart beat header.
+     *
+     * @param Frame $frame
+     * @return array
+     */
+    private function getHeartbeats(Frame $frame)
+    {
+        $beats = $frame['heart-beat'];
+        if ($beats) {
+            return explode(',', $beats, 2);
+        }
+        return [0, 0];
+    }
+
+    /**
+     * Indicates that a frame has been sent to the server.
+     *
+     * @param Frame $frame
+     * @return void
+     */
+    public function sentFrame(Frame $frame)
+    {
+        if ($this->enabled) {
+            $this->notifyBeat();
+            return;
+        }
+        if ($frame->getCommand() === Emitter::FRAME_CLIENT_CONNECT) {
+            $beats = $this->getHeartbeats($frame);
+            $this->intervalClient = $beats[0];
+            $this->notifyBeat();
```

Comment:
No, it should be possible to support the RabbitMQ "Implicit Connect" (https://www.rabbitmq.com/stomp.html) this works without a `CONNECT` frame - so the enabled flag is here to query the state, but not to control it. (see `testEmitterActivatedIfServerRequestsBeatsAndNoConnectFrameWasSend`)